src/databricks/sql/auth/auth.py

```diff
@@ -0,0 +1,103 @@
+# the Community Edition Terms of Service, (b) the Databricks Terms of
+# Service, or (c) another written agreement between Licensee and Databricks
+# for the use of the APIs.
+#
+# You may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from enum import Enum
+from databricks.sql.auth.authenticators import Authenticator, \
+    AccessTokenAuthenticator, UserPassAuthenticator, OAuthAuthenticator
+
+
+class AuthType(Enum):
+    DATABRICKS_OAUTH = "databricks-oauth"
+    # other supported types (access_token, user/pass) can be inferred
+    # we can add more types as needed later
+
+
+class ClientContext:
+    def __init__(self,
+                 hostname: str,
+                 username: str = None,
+                 password: str = None,
+                 token: str = None,
+                 auth_type: str = None,
+                 oauth_scopes: str = None,
+                 oauth_client_id: str = None,
+                 use_cert_as_auth: str = None,
+                 tls_client_cert_file: str = None):
+        self.hostname = hostname
+        self.username = username
+        self.password = password
+        self.token = token
+        self.auth_type = auth_type
+        self.oauth_scopes = oauth_scopes
+        self.oauth_client_id = oauth_client_id
+        self.use_cert_as_auth = use_cert_as_auth
+        self.tls_client_cert_file = tls_client_cert_file
+
+
+class SqlConnectorClientContext(ClientContext):
+    def __init__(self,
+                 hostname: str,
+                 username: str = None,
+                 password: str = None,
+                 access_token: str = None,
+                 auth_type: str = None,
+                 use_cert_as_auth: str = None,
+                 tls_client_cert_file: str = None):
+        super().__init__(oauth_scopes="sql offline_access",
+                         # to be changed once registered on the service side
+                         oauth_client_id="databricks-cli",
+                         hostname=hostname,
+                         username=username,
+                         password=password,
+                         token=access_token,
+                         auth_type=auth_type,
+                         use_cert_as_auth=use_cert_as_auth,
+                         tls_client_cert_file=tls_client_cert_file)
+
+
+def get_authenticator(cfg: ClientContext):
+    if cfg.auth_type == AuthType.DATABRICKS_OAUTH.value:
+        return OAuthAuthenticator(cfg.hostname, cfg.oauth_client_id, cfg.oauth_scopes)
+    elif cfg.token is not None:
```

Comment:
My concern with adding other authenticators in here too is it adds test burdens (which I cannot see tests being added in the same PR). I would leave the untested (out of scope) code out of the PRs too so we are sure that submitted code has been verified and has good test coverage.

Reply:
There are no tests at all at the moment. @moderakh Are you planning to include a couple in this PR?

If so, I recommend keeping these simpler authenticators around.

Reply:
Added tests covering the other auth providers.
https://github.com/databricks/databricks-sql-python/blob/03ab454bdd4d9a39d306873d916fb3496d4496df/tests/test_auth.py